python/hail/methods/qc.py

```diff
@@ -91,79 +93,117 @@ def sample_qc(dataset, name='sample_qc') -> MatrixTable:
+    field `dp_stats` is computed. If `mt` contains an entry field `GQ` of type
+    :py:data:`.tint32`, then the field `gq_stats` is computed. Both `dp_stats`
+    and `gq_stats` are structs with with four fields:
+
+    - `mean` (``float64``) -- Mean value.
+    - `stdev` (``float64``) -- Standard deviation (zero degrees of freedom).
+    - `min` (``int32``) -- Minimum value.
+    - `max` (``int32``) -- Maximum value.
+
+    If `mt` contains an entry field `GT` of type :py:data:`.tcall`, then the
+    following fields are computed:
+
+    - `AF` (``array<float64>``) -- Calculated allele frequency, one element
+      per allele, including the reference. Sums to one. Equivalent to
+      `AC` / `AN`.
+
+    - `AC` (``array<int32>``) -- Calculated allele count, one element per
+      allele, including the reference. Sums to `AN`.
+
+    - `AN` (``int32``) -- Total number of called alleles.
+
+    - `homozygote_count` (``array<int32>``) -- Number of homozygotes per
+      allele. One element per allele, including the reference.
+
+    - `n_called` (``int64``) -- Number of samples with a defined `GT`.
+    - `n_not_called` (``int64``) -- Number of samples with a missing `GT`.
+    - `call_rate` (``float32``) -- Fraction of samlpes with a defined `GT`.
+      Equivalent to `n_called` / :meth:`.count_cols`.
+    - `n_het` (``int64``) -- Number of heterozygous samples.
+    - `n_non_ref` (``int64``) -- Number of samples with at least one called
+      non-reference allele.
+    - `hwe` (``array<struct{r_expected_het_freq: float64, p_hwe: float64}>``) --
+      Hardy-Weinberg statistics, one element per alternate allele. Assumes
+      that all genotype calls are diploid, and treats each allele
+      independently.
 
     Parameters
     ----------
-    dataset : :class:`.MatrixTable`
+    mt : :class:`.MatrixTable`
         Dataset.
     name : :obj:`str`
         Name for resulting field.
 
     Returns
     -------
     :class:`.MatrixTable`
-        Dataset with a new row-indexed field `name`.
     """
-
-    return MatrixTable(Env.hail().methods.VariantQC.apply(require_biallelic(dataset, 'variant_qc')._jvds, name))
+    require_row_key_variant(mt, 'variant_qc')
+
+    exprs = {}
+    struct_exprs = []
+    fs = []
+
+    def has_field_of_type(name, dtype):
+        return name in mt.entry and mt[name].dtype == dtype
+
+    n_samples = mt.count_cols()
+
+    if has_field_of_type('DP', hl.tint32):
+        exprs['dp_stats'] = hl.agg.stats(mt.DP).select('mean', 'stdev', 'min', 'max')
+
+    if has_field_of_type('GQ', hl.tint32):
+        exprs['gq_stats'] = hl.agg.stats(mt.GQ).select('mean', 'stdev', 'min', 'max')
+
+    if has_field_of_type('GT',  hl.tcall):
+        exprs['n_called'] = hl.agg.count_where(hl.is_defined(mt['GT']))
+        struct_exprs.append(hl.agg.call_stats(mt.GT, mt.alleles))
+        fs.append(identity)
+
+    # the structure of this function makes it easy to add new nested computations
+    def make_struct(*struct_args):
```

Comment:
Can you move this to a utils file? I'm sure this will be useful for methods developers in the future. I see it depends on other variables outside the function though.

Reply:
I can't, because it references a local variable above it. I can't add that as an argument, or `bind` won't work.